node/lib/util/merge_bare_util.js

```diff
@@ -0,0 +1,287 @@
+        errorMessage += `Conflicting entries for submodule: ` + 
+            `${colors.red(name)}\n`;
+    }
+    errorMessage += "Automatic merge failed\n";
+    return errorMessage;
+}
+
+/**
+ * Merge  `theirCommit` into `ourCommit` in the specified `repo` with specific
+ * commitMessage. Return `null` if there are merge conflicts and the conflicts
+ * cannot be resolved automatically. Return an object describing merge commits
+ * otherwise. Use our commit as the merged commit if our commit is up to date,
+ * use theirs if this is a fast forward merge and return a new merge commit
+ * otherwise. Throw a `UserError` if there are no commits in common between 
+ * `theirCommit` and `ourCommit`.
+ *
+ * @async
+ * @param {NodeGit.Repository} repo
+ * @param {NodeGit.Commit}     ourCommit
+ * @param {NodeGit.Commit}     theirCommit
+ * @param {String}             commitMessage
+ * @return {Object}
+ * @return {String|null} return.metaCommit
+ * @return {Object}      return.submoduleCommits  map from submodule to commit
+ * @return {String|null} return.errorMessage
+ */
+exports.merge = co.wrap(function *(repo,
+                                   ourCommit,
+                                   theirCommit,
+                                   commitMessage) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(ourCommit, NodeGit.Commit);
+    assert.instanceOf(theirCommit, NodeGit.Commit);
+    assert.isString(commitMessage);
+
+    const baseCommit = yield GitUtil.getMergeBase(repo, ourCommit, theirCommit);
+
+    if (null === baseCommit) {
+        throw new UserError(`No commits in common with `+
+            `${colors.red(GitUtil.shortSha(ourCommit.id().tostrS()))} and ` +
+            `${colors.red(GitUtil.shortSha(theirCommit.id().tostrS()))}`);
+    }
+
+    yield CherryPickUtil.ensureNoURLChanges(repo, ourCommit, theirCommit);
+
+    const result = {
+        metaCommit: null,
+        submoduleCommits: {},
+        errorMessage: null,
+    };
+
+    const ourCommitSha = ourCommit.id().tostrS();
+    const theirCommitSha = theirCommit.id().tostrS();
+    if (ourCommitSha === theirCommitSha) {
+        console.log(`Nothing to do for merging ${colors.green(ourCommitSha)}` +
+            `into itself.`);
+        result.metaCommit = ourCommitSha;
+        return result;
+    }
+
+    const upToDate  = yield NodeGit.Graph.descendantOf(repo,
+                                                       ourCommitSha,
+                                                       theirCommitSha);
+
+    if (upToDate) {
+        console.log(`${colors.green(ourCommitSha)} is up-to-date.`);
+        result.metaCommit = ourCommitSha;
+        return result;                                                // RETURN
+    }
+
+    const canFF  = yield NodeGit.Graph.descendantOf(repo,
+                                                    theirCommitSha,
+                                                    ourCommitSha);
+
+    if (canFF) {
```

Comment:
Wait, I thought we were not doing fast-forward merges?  For the work we are doing now, non-ff is better.  Can you provide an option to turn ff off?